plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/FileSystemExchangeSource.java

```diff
@@ -14,121 +14,168 @@
+    @Override
+    public synchronized void addSourceHandles(List<ExchangeSourceHandle> handles)
+    {
+        if (closed.get()) {
+            return;
         }
-        this.readers = readers.build();
+        files.addAll(getFiles(handles));
+        createReadersIfNecessary();
+    }
+
+    @Override
+    public synchronized void noMoreSourceHandles()
+    {
+        noMoreFiles = true;
+        createReadersIfNecessary();
     }
 
     @Override
     public CompletableFuture<Void> isBlocked()
     {
-        CompletableFuture<Void> blocked = this.blocked;
+        if (closed.get()) {
+            return NOT_BLOCKED;
+        }
+
+        CompletableFuture<Void> blocked = this.blocked.get();
         if (blocked != null && !blocked.isDone()) {
             return blocked;
         }
-        for (ExchangeStorageReader reader : readers) {
+
+        List<ExchangeStorageReader> readers = this.readers.get();
+        // regular loop for efficiency
+        for (int i = 0; i < readers.size(); i++) {
+            ExchangeStorageReader reader = readers.get(i);
             if (reader.isBlocked().isDone()) {
                 return NOT_BLOCKED;
             }
         }
+
         synchronized (this) {
-            if (this.blocked == null || this.blocked.isDone()) {
-                this.blocked = stats.getExchangeSourceBlocked().record(toCompletableFuture(
+            if (!blockedOnSourceHandles.isDone()) {
+                blocked = toCompletableFuture(nonCancellationPropagating(blockedOnSourceHandles));
+            }
+            else if (readers.isEmpty()) {
+                blocked = NOT_BLOCKED;
+            }
+            else {
+                blocked = toCompletableFuture(
                         nonCancellationPropagating(
                                 whenAnyComplete(readers.stream()
                                         .map(ExchangeStorageReader::isBlocked)
-                                        .collect(toImmutableList())))));
+                                        .collect(toImmutableList()))));
             }
-            return this.blocked;
+            blocked = stats.getExchangeSourceBlocked().record(blocked);
+            this.blocked.set(blocked);
+            return blocked;
         }
     }
 
     @Override
     public boolean isFinished()
     {
-        if (closed) {
-            return true;
-        }
-
-        for (ExchangeStorageReader reader : readers) {
-            if (!reader.isFinished()) {
-                return false;
-            }
-        }
-        return true;
+        return closed.get();
     }
 
     @Nullable
     @Override
     public Slice read()
     {
-        if (closed) {
+        if (closed.get()) {
             return null;
         }
 
-        for (ExchangeStorageReader reader : readers) {
+        Slice data = null;
+        List<ExchangeStorageReader> readers = this.readers.get();
+        // regular loop for efficiency
+        for (int i = 0; i < readers.size(); i++) {
```

Comment:
what is wrong with `for (ExchangeStorageReader reader : readers) {`
Is it really less efficient?

Reply:
The `for (ExchangeStorageReader reader : readers)` creates an iterator. The `for (int i = 0; i < readers.size(); i++) {` variant is more efficient: https://howtodoinjava.com/java/collections/performance-comparison-looping-a-list/